kv: remove dependency on storage

RaduBerinde opened the pull request:
Removing the kv dependency on storage by moving most of LocalTestCluster to a
`testutils` package. The kv-specific part stays behind as a callback.

Also had to make a kv copy of DefaultHeartbeatInterval; added a test to verify
the value is the same.

Fixes #6246.

Second commit is a test for the dependency.

Current dependency graph (excluding util,keys,rpc,cliflags,base,security,roachpb): http://postimg.org/image/o9tbeg5y9/

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6466)

<!-- Reviewable:end -->

bdarnell commented:
LGTM

---

Reviewed 9 of 9 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[kv/local_test_cluster_util.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6466#-KGseNs-nRNxQjVT7PL3:-KGseNs-nRNxQjVT7PL4:-754994258) ([raw file](https://github.com/cockroachdb/cockroach/blob/88461fb68bf42e30f7fe1a22c35790b9cb5464e2/kv/local_test_cluster_util.go#L76)):*
There's very little unexported here. Is it just this `nodeDescriptor` field, the `rpcSendFn` type used above, and the default values? We should probably just export those instead of leaving this function behind in package `kv`. (`rpcSendFn`is already the type of an exported field in `DistSenderContext` so it really ought to be exported)

---

*[kv/txn_coord_sender.go, line 48 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6466#-KGsez28FJ29Idf2XzH0:-KGsez29adXhzAbtrxP_:1725806186) ([raw file](https://github.com/cockroachdb/cockroach/blob/88461fb68bf42e30f7fe1a22c35790b9cb5464e2/kv/txn_coord_sender.go#L48)):*
Move this to package `base` instead of duplicating it.

---

*[testutils/localtestcluster/local_test_cluster.go, line 41 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6466#-KGsdPPrE73t1hkB7_6r:-KGsdPPrE73t1hkB7_6s:-1602661243) ([raw file](https://github.com/cockroachdb/cockroach/blob/88461fb68bf42e30f7fe1a22c35790b9cb5464e2/testutils/localtestcluster/local_test_cluster.go#L41)):*
Update this example for the new argument.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6466#-:-KGsfQxIIjlQU8-9nR59:676834233)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[kv/local_test_cluster_util.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6466#-KGseNs-nRNxQjVT7PL3:-KGsguY7xxDPRku1WEFv:1041496505) ([raw file](https://github.com/cockroachdb/cockroach/blob/88461fb68bf42e30f7fe1a22c35790b9cb5464e2/kv/local_test_cluster_util.go#L76)):*
Exporting those and keeping the code in one place is the first thing I tried. The problem is that `testutils/localtestcluster` cannot depend on `kv` or we get a circular dependency when we try to use it from `kv` tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6466)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[kv/local_test_cluster_util.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6466#-KGseNs-nRNxQjVT7PL3:-KGsiNegTHK4Itu01SF3:-1509255054) ([raw file](https://github.com/cockroachdb/cockroach/blob/88461fb68bf42e30f7fe1a22c35790b9cb5464e2/kv/local_test_cluster_util.go#L76)):*
Ah, so it's not just dependencies on unexported state, its' dependency on `kv` at all. Got it. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6466#-:-KGsiZ4r9h72tozE8pGQ:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
kv: add test to check for kv->storage dependency

RaduBerinde committed:
kv: remove dependency on storage

Removing the kv dependency on storage by moving most of LocalTestCluster to a
`testutils` package. The kv-specific part stays behind as a callback.

Also had to make a kv copy of DefaultHeartbeatInterval; added a test to verify
the value is the same.

Fixes #6246.

RaduBerinde commented:
TFTR! Updated.

---

Review status: 6 of 18 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6466#-:-KGsyY_U7H7-bn7DzFkL:-1512379193)*
<!-- Sent from Reviewable.io -->
